contracts/src/AgentExecutor.sol

```diff
@@ -36,4 +50,25 @@ contract AgentExecutor {
     function _transferToken(address token, address recipient, uint128 amount) internal {
         IERC20(token).safeTransfer(recipient, amount);
     }
+
+    /// @dev Register native asset from polkadto as ERC20 `token`.
+    function _registerToken(bytes32 agentID, bytes32 tokenID, string memory name, string memory symbol, uint8 decimals)
+        internal
+    {
+        IERC20 token = new ERC20(name, symbol, decimals);
+        Gateway(msg.sender).registerForeignToken(tokenID, address(token), agentID);
```

Comment:
The call stack of registering a token is:
```
Gateway.submitV1 -> AgentExecutor.execute -> AgentExecutor._registerToken -> Gateway. registerForeignToken
```
From what I can see this is done to make sure the owner of ERC20 token is the agent. This can be simplified if we add a `setOwner` method to `ERC20`. This will allow the call stack to be:
```
Gateway.submitV1 -> Gateway.registerForeignToken
```
Where `Gateway.registerForeignToken` creates a new ERC20 to which initially it is the owner, and then assigns ownership to the agent using the `setOwner` function.

This will also allow agentId to removed from the `execute` function signature. This will also probably save gas.

Reply:
Yeah, I agree that the current flow is overly complicated. We should explore alternatives like the one you suggest.

Though instead of the Gateway passing ownership of the ERC20 token to the Agent, the gateway calls something like:

```solidity
Agent(agent).invoke(AGENT_EXECUTOR, AgentExecutor.createERC20(params...))
```

Reply:
Vincent, is that something https://github.com/Snowfork/snowbridge/pull/1155#discussion_r1541966351 we tried and then reverted?